### PR TITLE
gpt-oss: stop dropping the causal mask during compiled inference

### LINKED ISSUE #1209: gpt-oss: torch.compile silently corrupts output on B200 (ppl 4.5 -> 643 MXFP4, 4.9 -> 73 bnb-4bit)
Found while measuring #1208. Not caused by it, and not specific to MXFP4.

## What happens

`unsloth/gpt-oss-20b` on a B200 gives correct output with `UNSLOTH_COMPILE_DISABLE=1` and badly degraded output without it. Perplexity on 163 tokens of fixed prose, greedy, same weights, same process setup, fresh compiled cache each time:

| path | compile disabled | compile enabled |
|---|---|---|
| native MXFP4 | 4.507 | 642.7 |
| `unsloth/gpt-oss-20b-unsloth-bnb-4bit` | 4.936 | 73.3 |

Next-token top-1 against the text drops from 64.4% to 6.7% on the MXFP4 path and from 62.7% to 18.7% on bnb-4bit. The model stays fluent, so short prompts still answer correctly ("capital of France" is still Paris) and only measurement shows the damage.

The bf16 reference on the same host is perplexity 4.487, so the compile-disabled numbers are right and the compile-enabled ones are wrong.

## Where it starts

Per-module forward hooks on both runs, same input, comparing outputs layer by layer:

```
model.layers.0.self_attn   | GptOssAttention        max|diff| 14.25   rel 0.63
model.layers.0.mlp.experts | Mxfp4GptOssExperts     max|diff| 33.50   rel 0.39
model.layers.0.mlp         | GptOssMLP              max|diff| 33.50   rel 0.39
```

So it is already wrong in the first decoder layer's attention, before the MoE, and grows from there (by layer 16 the decoder layer output differs by rel 1.0). The bound expert forward is the same object in both runs (`forward_mxfp4_gpt_oss_with_lora`, using `matmul_ogs`), so this is not a case of the wrong forward being selected.

`unsloth_compiled_cache/` was deleted before the compiled run, so this is not stale-cache carryover. The flavor marker `.unsloth_gpt_oss_compiled_flavor` reads `stock` for a native MXFP4 load, which may or may not be related.

## Environment

- B200, sm_100, driver CUDA 13.0
- torch 2.12.1+cu130, triton 3.7.1
- transformers 4.56.2, trl 0.22.2, unsloth 2026.9.4, unsloth_zoo 2026.9.3
- python 3.12, Linux

The same notebook on Kaggle T4x2 (sm_75) with compile enabled answers short questions correctly; I have not measured perplexity there, so I cannot say yet whether this is Blackwell-specific or general. That is the first thing to check.

## Repro

```python
import os
os.environ["UNSLOTH_COMPILE_DISABLE"] = "1"   # flip to compare
import torch
from unsloth import FastLanguageModel

model, tok = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b-unsloth-bnb-4bit",
    dtype = None, max_seq_length = 4096, load_in_4bit = True, full_finetuning = False,
)
ids = tok("The Apollo program was the third United States human spaceflight program "
          "carried out by the National Aeronautics and Space Administration.",
          return_tensors = "pt").input_ids.to("cuda")
with torch.inference_mode():
    logits = model(ids).logits.float()
nll = torch.nn.functional.cross_entropy(logits[0, :-1], ids[0, 1:])
print("perplexity", nll.exp().item())
```

### PR DESCRIPTION
**Pull request by danielhanchen**
Fixes #1209.

## What happens

With `torch.compile` on (the default), `unsloth/gpt-oss-20b` produces fluent but badly degraded output. Perplexity on 163 tokens of fixed prose, B200, same weights, fresh compiled cache:

| path | compile disabled | compile enabled (before) | compile enabled (this PR) |
|---|---|---|---|
| native MXFP4 | 4.507 | 642.732 | **4.500** |
| `gpt-oss-20b-unsloth-bnb-4bit` | 4.936 | 73.301 | **4.959** |

Against a stock-transformers bf16 reference (perplexity 4.487) on the same host, top-1 agreement goes from 10.4% to **97.5%** and mean KL from 4.65 to **0.0014**.

Because the model stays fluent, short prompts still answer correctly ("capital of France" is still Paris) and nothing crashes. Only a perplexity measurement exposes it.

## Why

`patch_GptOssModel` wraps `create_causal_mask` so that flex-attention training can skip building a dense mask: flex builds its own `BlockMask` and a dense one is an O(seq_len^2) allocation that OOMs at long context. The skip fired on `input_embeds.requires_grad` alone:

```python
if input_embeds is not None and input_embeds.requires_grad:
    if "attention_mask" in kwargs:
        return kwargs["attention_mask"]
```

`requires_grad` is not a training signal here. unsloth calls `enable_input_require_grads()` on every LoRA-capable load, so the embedding output requires grad during inference too. Measured, both compiled and uncompiled: `embedding output requires_grad: True`, `model.training: False`.

That wrapper was only ever safe in company. Zoo's own `GptOssModel.forward` clears the flag for inference before the factories run:

```python
if not self.training:
    inputs_embeds.requires_grad_(False)
```

On transformers 4.56.2 that forward does not install. `patch_function` logs `Unsloth: Skipped GptOssModel.forward`, so stock `GptOssModel.forward` runs, calls the wrapped module-level `create_causal_mask`, the skip fires, and eager attention receives `attention_mask=None`, which for eager means no causal masking at all.

Traced directly, one `create_causal_mask` call per forward:

| | compile off | compile on |
|---|---|---|
| `create_causal_mask` is | stock transformers | zoo's `return_attention_mask` |
| `input_embeds.requires_grad` | True | True |
| returns | `Tensor` | **`None`** |
| mask reaching `GptOssAttention.forward` | `(1, 1, 12, 12)` | **`None`** |

`UNSLOTH_COMPILE_DISABLE=1` makes `patch_GptOssModel` return early at `gpt_oss.py:2563`, which is exactly why the uncompiled runs were correct and hid this.

## The fix

The skip exists because `flex_attention_with_sink` builds its own `BlockMask` and a dense
mask costs O(seq_len^2). The patched `GptOssAttention.forward` selects that path from
`self.training` alone (`gpt_oss.py:2319`), not from `_attn_implementation`, so the gate has
to key on the same thing:

- `GptOssModel.forward` records its own `self.training` on the config as
  `_unsloth_gpt_oss_model_training`, installed after the forward replacement so it wraps
  whichever forward ends up live (zoo's own, or stock when the signature does not match)
  and idempotent across the three `TEMPORARY_PATCHES` passes;
- `patch_GptOssAttention` records whether the flex-routing forward actually installed;
- the skip is taken only when the model is training and that forward is live. Where no
  model forward recorded a flag, it falls back to a grad-shaped guess, which still excludes
  `no_grad` inference.

Everything else falls through to building the dense mask, which is what eager needs.

Nothing else in either repo used `requires_grad` as a training proxy for masks: `misc.py:214`
pairs it with `is_leaf` for gradient checkpointing, and `unsloth/models/llama.py:964-993`
toggles it deliberately around checkpointing.

## Not a torch or Blackwell bug

Same host, same torch 2.12.1+cu130: `unsloth/Llama-3.2-1B-Instruct` and `unsloth/Qwen3-1.7B` are **bit-identical** compiled vs uncompiled (NLL 1.7423 and 1.5789 to four decimals both ways). Only gpt-oss goes through this wrapper.

Per-module forward hooks put the first divergence at `model.layers.0.self_attn` (rel 0.63), before the MoE, growing to rel 1.0 by layer 16. Every compiled helper was cleared individually first: rotary, `apply_rotary_pos_emb`, `repeat_kv`, RMSNorm and the router are bit-identical compiled vs eager on real weights; `eager_attention_forward` differs only by bf16 rounding (rel 0.005).

## Tests

`tests/test_gpt_oss_causal_mask_not_skipped.py`, 11 CPU-only tests driving the real installed
wrapper: eager inference with `requires_grad=True` must get a mask (the regression), eager
inference under `no_grad`, training with flex still takes the skip, training with the patched
attention live takes the skip even on an eager `_attn_implementation`, training without the
patched attention builds a mask, flex under `no_grad` builds a mask, a `BlockMask` case, the
config swap to eager and its restore, a call with no config, the no-flag-recorded fallback,
and positional-argument calls.

Measured on a B200, transformers 5.17.0, `gpt-oss-20b-unsloth-bnb-4bit`:

| | loss | dense masks built | peak memory |
|---|---|---|---|
| zoo main | 2.469727 | 0 | 16.62 GiB |
| this PR | 2.460861 | 0 | 16.62 GiB |

(zoo's own `GptOssModel.forward` installs on 5.17.0 and does not call the mask factories at
all, so neither side builds one; on 4.56.2, where the forward is skipped, both sides return
the caller's 2D mask.)

Inference on the same host stays fixed: perplexity 5.594 against the bf16 reference 4.487,
89.6% top-1 agreement, coherent generations. Earlier, under transformers 4.56.2 and native
MXFP4, the same measurement was 642.7 before the fix and 4.500 after (bnb-4bit 73.3 to 4.959).

Cross-platform staging CI runs Ubuntu, Windows and Apple Silicon macOS off the org queue.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Training measurements, expanded, because the first pass compared one run against one run and the difference looked meaningful when it was not.

Two LoRA steps on `unsloth/gpt-oss-20b-unsloth-bnb-4bit`, compile on, seed 3407, fresh compiled cache each run, two runs per variant:

| variant | loss | peak memory | mask the wrapper returns |
|---|---|---|---|
| before | 2.549004 | 18.93 GiB | `tensor(2, 128) int64` |
| before | 2.544762 | 18.93 GiB | `tensor(2, 128) int64` |
| after | 2.549896 | 18.93 GiB | `tensor(2, 1, 128, 128) bfloat16` |
| after | 2.541159 | 18.93 GiB | `tensor(2, 1, 128, 128) bfloat16` |

Before-to-before spread is 0.0042 and after-to-after is 0.0087, so the two variants are indistinguishable at two steps. Peak memory is identical.

What does change is the mask handed to `GptOssAttention.forward` during training: a 2D `int64` padding mask before, a 4D `bfloat16` mask with actual causal structure after (verified by checking that the last query row is not uniform across keys).

This configuration reports `_attn_implementation = eager` with zoo's own `patch_GptOssAttention.<locals>.forward` bound, which evidently tolerates the 2D mask, so I am **not** claiming training was previously wrong. Two steps cannot show that either way. The inference claim is the one this PR rests on, and that one is unambiguous: perplexity 642.7 to 4.5.

Worth a longer training comparison by someone with a real gpt-oss fine-tune to hand, since the mask now reaching attention is a different object.

File: unsloth_zoo/temporary_patches/gpt_oss.py [hunk 1]
```diff
@@ -2433,7 +2442,21 @@ def return_attention_mask(*args, **kwargs):
...
+            # inference too, and skipping there hands eager attention a None mask,
+            # i.e. no causal masking at all -- fluent but wrong output.
+            _config = _find_config(args, kwargs)
+            _is_flex = getattr(_config, "_attn_implementation", None) == "flex_attention"
+
+            if (
+                _is_flex
+                and torch.is_grad_enabled()
+                and input_embeds is not None
+                and input_embeds.requires_grad
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gpt_oss.py]: Correct, and confirmed against the patched attention forward: it selects flex_attention_with_sink from self.training alone (gpt_oss.py:2319), so an eager _attn_implementation during training still routes to flex and the dense mask I was building would have gone unused.

Fixed in ace55a86. GptOssModel.forward now records its own training state on the config (`_unsloth_gpt_oss_model_training`), patch_GptOssAttention records whether the flex-routing forward actually installed, and the skip is taken when both hold rather than keying on _attn_implementation. The recorder is installed after the GptOssModel.forward replacement so it wraps whichever forward ends up live (zoo's own, or stock when the signature does not match), and it is idempotent across the three TEMPORARY_PATCHES passes: wrapper nesting depth stays 1 over three runs. Where no model forward recorded a flag, it falls back to the grad-shaped guess, which still excludes no_grad inference.

Verified on a B200 with transformers 5.17.0 and unsloth/gpt-oss-20b-unsloth-bnb-4bit:

- training A/B against zoo main is unchanged: loss 2.4697 (main) vs 2.4609 (fixed), peak 16.62 GiB in both, and zero dense masks built in either;
- inference perplexity stays fixed at 5.594 against the bf16 reference 4.487, 89.6% top-1 agreement, coherent generations;
- unit tests are now 11, covering training-plus-flex, training without the patched attention, inference with requires_grad set, the BlockMask case and the no-flag-recorded fallback.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

File: unsloth_zoo/temporary_patches/gpt_oss.py [hunk 2]
```diff
@@ -2433,7 +2475,25 @@ def return_attention_mask(*args, **kwargs):
...
+            _training = getattr(_config, _TRAINING_FLAG_ATTR, None)
+            if _training is None:
+                # Unusual caller, no model forward recorded it: at least exclude
+                # no_grad inference.
+                _training = bool(
+                    torch.is_grad_enabled()
+                    and input_embeds is not None
+                    and input_embeds.requires_grad
+                )
+            _flex_will_run = _is_flex or _GPT_OSS_FLEX_SINK_ATTENTION_INSTALLED
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gpt_oss.py]: Agreed, and confirmed against the installed transformers. `flex_attention_forward` (transformers/integrations/flex_attention.py:262) takes causality entirely from the `attention_mask` it is handed: a `BlockMask` becomes `block_mask`, anything else becomes `score_mask` and is indexed `score_mask[batch_idx][0][q_idx][kv_idx]`. It builds nothing of its own, so with the stock forward a flex `_attn_implementation` alone is not a reason to skip; only zoo's patched forward constructs its own BlockMask.

Fixed in 8b52ecc7: the skip now requires `_GPT_OSS_FLEX_SINK_ATTENTION_INSTALLED`, and `_is_flex` is used only for the existing swap to eager while the dense mask is built. A flex config without the patched forward now falls through to the factory, which is what that path needs. Added `test_flex_training_builds_the_mask_when_the_flex_forward_is_absent` to hold it; the file is at 12 tests.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gpt_oss.py [hunk 3]
```diff
@@ -2433,7 +2475,26 @@ def return_attention_mask(*args, **kwargs):
...
+                # no_grad inference.
+                _training = bool(
+                    torch.is_grad_enabled()
+                    and input_embeds is not None
+                    and input_embeds.requires_grad
+                )
+            # Only zoo's patched forward builds its own BlockMask. Stock transformers
+            # flex attention takes causality from whatever this factory returns, so a
+            # flex _attn_implementation on its own is not a reason to skip.
+            if _training and _GPT_OSS_FLEX_SINK_ATTENTION_INSTALLED:
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gpt_oss.py]: Genuine, and the same class of bug one level up. zoo's own `GptOssModel.forward` sets `attention_mask = None` for `self.training` on the same assumption that flex_attention_with_sink rebuilds it, so with the attention patch absent the mask I now build gets dropped again before the stock attention layers see it.

Fixed in b03a1e1d: that branch is now `if self.training and _GPT_OSS_FLEX_SINK_ATTENTION_INSTALLED`. The flag resolves from the patch module's globals at call time (the installed forward's `__globals__` carries it, verified), same as the other module-level names that forward already reads.

Locked with `test_model_forward_drops_the_mask_only_when_the_flex_forward_is_installed`, which parses `patch_GptOssModel` and fails on any `attention_mask = None` whose enclosing condition does not mention the flag, so a later edit cannot quietly reintroduce the bare `if self.training`. The file is at 13 tests.

Re-measured on a B200, transformers 5.17.0, `gpt-oss-20b-unsloth-bnb-4bit`: training loss 2.4689 at 16.62 GiB peak, against 2.4697 at 16.62 GiB for zoo main, and inference perplexity still 5.594 against the bf16 reference 4.487.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: tests/test_gpt_oss_causal_mask_not_skipped.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+                 "_old_create_sliding_window_causal_mask"):
+        if hasattr(masking_utils, name):
+            monkeypatch.delattr(masking_utils, name, raising = False)
+    monkeypatch.setattr(masking_utils, "create_causal_mask", factory, raising = False)
+    monkeypatch.setattr(masking_utils, "create_sliding_window_causal_mask", factory, raising = False)
+    monkeypatch.setattr(masking_utils, "create_masks_for_generate", factory, raising = False)
+    import transformers.generation.utils as generation_utils
+    monkeypatch.setattr(generation_utils, "create_masks_for_generate", factory, raising = False)
+
+    gpt_oss.patch_GptOssModel()
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gpt_oss_causal_mask_not_skipped.py]: Correct. `patch_GptOssModel` writes nine bindings across three modules plus `GptOssModel.forward`, and the fixture only registered four of them with monkeypatch, so `modeling_gpt_oss.create_causal_mask`, its sliding-window twin and the patched `forward` were left wrapping this fixture's `factory`.

Fixed in a0acbf3d: the fixture is now a yield fixture that snapshots every binding the patch touches, including the names it creates (`__patched_causal_mask__`, `_old_create_causal_mask`, `_old_create_sliding_window_causal_mask`), and restores or deletes each one in a `finally`. `GptOssModel.forward` is read from `__dict__` rather than with `getattr`, since getattr would return an inherited forward and restoring it would define one on the class where there was none.

Verified in-process: snapshot every binding after collection, run the suite, compare identities afterwards. Before the fix the model-module bindings and `forward` came back changed; after it the diff is empty and no sentinel names survive.

The branch is also rebased onto main now that #1208 has merged (no conflicts).
